exchange/auction_test.go

```diff
@@ -94,92 +112,143 @@ func loadCacheSpec(filename string) (*cacheSpec, error) {
-			if !vast {
-				// not testing VAST XML for custom cache key test
-				eq := jsonpatch.Equal(cExpected.Data, cFound.Data)
-				if !eq {
-					continue
-				}
-			}
+	if len(specData.ExpectedCacheables) > len(cache.items) {
+		//t.Errorf("%s:  Less elements were cached than expected \n", fileDisplayName)
+		t.Errorf("%s:  [CACHE_ERROR] Less elements were cached than expected \n", fileDisplayName)
+	} else if len(specData.ExpectedCacheables) < len(cache.items) {
+		t.Errorf("%s:  [CACHE_ERROR] More elements were cached than expected \n", fileDisplayName)
+	} else { // len(specData.ExpectedCacheables) == len(cache.items)
+		var ht map[string]*cacheComparator = make(map[string]*cacheComparator)
+		var formattedData string
+		var compareString string
 
-			// make sure Key value is as expected
-			keymatch := false
-			if len(cExpected.Key) > 0 {
-				// can only verify prefix; remainder is random
-				if strings.HasPrefix(cFound.Key, cExpected.Key) {
-					keymatch = true
-				}
-			} else {
-				if len(cFound.Key) == 0 {
-					// Key is expected to be empty
-					keymatch = true
-				}
+		for _, cExpected := range specData.ExpectedCacheables {
+			formattedData = strings.Replace(string(cExpected.Data), "\\", "", -1)
+			formattedData = strings.Replace(formattedData, " ", "", -1)
+			compareString = fmt.Sprintf("%s_%d", formattedData, cExpected.TTLSeconds)
+			// init cacheComparator element before hashing it. If needed
+			if _, ok := ht[compareString]; !ok {
+				ht[compareString] = &cacheComparator{0, make([]string, 0), make([]string, 0)}
 			}
-			if !keymatch {
-				continue
+			ht[compareString].freq += 1
+			if targData.includeCacheVast {
+				ht[compareString].expectedKeys = append(ht[compareString].expectedKeys, cExpected.Key)
 			}
-
-			// make sure TTLSeconds section matches exactly
-			if cExpected.TTLSeconds == cFound.TTLSeconds {
-				found++
+		}
+		for _, cFound := range cache.items {
+			formattedData = strings.Replace(string(cFound.Data), "\\", "", -1)
+			formattedData = strings.Replace(formattedData, " ", "", -1)
+			compareString = fmt.Sprintf("\"%s\"_%d", formattedData, cFound.TTLSeconds)
+			// init cacheComparator element before hashing it. If needed
+			if _, ok := ht[compareString]; !ok {
+				ht[compareString] = &cacheComparator{0, make([]string, 0), make([]string, 0)}
+			}
+			ht[compareString].freq -= 1
+			if targData.includeCacheVast {
+				ht[compareString].actualKeys = append(ht[compareString].actualKeys, cFound.Key)
+			}
+		}
+		for k, cachedElements := range ht {
+			if cachedElements.freq > 0 {
+				t.Errorf("%s:  [CACHE_ERROR] Cache inconsistensy. Element %s was not expected to get cached\n", fileDisplayName, k)
+			} else if cachedElements.freq < 0 {
+				t.Errorf("%s:  [CACHE_ERROR] Cache inconsistensy. We cached some more elements (%s) than expected.\n", fileDisplayName, k)
+			} else if targData.includeCacheVast {
+				for i := 0; i < cachedElements.freq; i++ {
```

Comment:
My comment from the previous review disappeared!  It stated that this code is never reached because cachedElements.freq is necessarily 0 if we're in this else statement.  Probably the length of expected keys is the right counter to loop on.

Reply:
It is, good call.